modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+}
+
+// func (t *tenantUsage) getSeries(labels, values []string, maxCardinality uint64) *bucket {
+func (t *tenantUsage) getSeries(buffer []string, maxCardinality uint64) *bucket {
+	h := hash(t.sortedKeys, buffer)
+
+	b := t.series[h]
+	if b == nil {
+		// Before creating a new series, check for cardinality limit.
+		if uint64(len(t.series)) >= maxCardinality {
+			// Overflow
+			// This tenant is at the maximum number of series.  In this case all data
+			// goes into the final overflow bucket. It has the same dimensions as the
+			// current configuration, except every label is overridden to the special overflow value.
+			for k := range buffer {
+				buffer[k] = overflowLabel
+			}
+			h = t.overflow
+			b = t.series[h]
+		}
+	}
+
+	if b == nil {
+		// First encounter with this series. Initialize it.
+		// Detach a copy of the values
+		v := make([]string, len(buffer))
+		copy(v, buffer)
+		b = &bucket{
+			// Metric description - constant for this pass now that the dimensions are known
+			descr:  prometheus.NewDesc("tempo_usage_tracker_bytes_received_total", "bytes total received with these attributes", t.sortedKeys, t.constLabels),
+			labels: v,
+		}
+		t.series[h] = b
+	}
+	return b
+}
+
+type Tracker struct {
+	mtx      sync.Mutex
+	name     string
+	tenants  map[string]*tenantUsage
+	labelsFn tenantLabelsFunc
+	maxFn    tenantMaxFunc
+	reg      *prometheus.Registry
+	cfg      PerTrackerConfig
+}
+
+func NewTracker(cfg PerTrackerConfig, name string, labelsFn tenantLabelsFunc, maxFn tenantMaxFunc) (*Tracker, error) {
+	u := &Tracker{
+		cfg:      cfg,
+		name:     name,
+		tenants:  make(map[string]*tenantUsage),
+		labelsFn: labelsFn,
+		maxFn:    maxFn,
+		reg:      prometheus.NewRegistry(),
+	}
+
+	err := u.reg.Register(u)
+	if err != nil {
+		return nil, err
+	}
+
+	go u.PurgeRoutine()
+
+	return u, nil
+}
+
+// getTenant must be called under lock.
+func (u *Tracker) getTenant(tenant string) *tenantUsage {
+	data := u.tenants[tenant]
+	if data == nil {
+		data = &tenantUsage{
+			series: make(map[uint64]*bucket),
+			constLabels: prometheus.Labels{
+				tenantLabel:  tenant,
+				trackerLabel: u.name,
+			},
+		}
+		u.tenants[tenant] = data
+	}
+	return data
+}
+
+func (u *Tracker) Observe(tenant string, batches []*v1.ResourceSpans) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	dimensions := u.labelsFn(tenant)
+	if len(dimensions) == 0 {
+		// Not configured
+		// TODO - Should we put it all in the unattributed bucket instead?
+		return
+	}
+
+	max := u.maxFn(tenant)
+	if max == 0 {
+		max = u.cfg.MaxCardinality
+	}
+
+	var (
+		now                             = time.Now().Unix()
+		data                            = u.getTenant(tenant)
+		mapping, buffer1, buffer2, last = data.GetBuffersForDimensions(dimensions)
+	)
+
+	for _, batch := range batches {
+		unaccountedForBatchData, totalSpanCount := nonSpanDataLength(batch)
+
+		if totalSpanCount == 0 {
+			// Mainly to prevent a panic below, but is this even possible?
```

Comment:
i think technically yes, so this is good guard code. i'm imagining someone writing a custom client to push to tempo or perhaps a bug in one of the otel clients creating this situation.